### ISSUE TITLE
Managing a whitelist

### ISSUE THREAD

**Issue by markmacgillivray** [CottageLabs/webber_server#10]
Server must provide a whitelist text file at a whitelist URL, so that the plugins (and users via web browser) can GET the whitelist and read it.

Server must also provide an API for updating the whitelist - most likely best option is for admin users to maintain the text file themselves somewhere, then just POST/PUT updated copies of it to an API address that will accept their new list as content if they authenticate as admin, and then replaces the file content with the new content provided.

Consider if this is enough, or if we need the API to support adding / removing specific URL lines to the file (probably not, but keep it in mind). The problem with this is that the server version MUST then be the canonical version - a subsequent upload by an administrator with a not up to date version would overwrite explicit additions made by any other admin user interaction. So would prefer to keep the canonical version off-server, wherever admin users prefer (e.g. their local shared network or wherever) then they just POST whole new file. It is very unlikely that such a file is going to ever reach a very large size such that this method would become problematic.

Consider if it is worth validating the POST of a new list and rejecting with errors if it does not conform. In this case, the conformity should be to allow blank lines, allow any line starting with a # as a comment / section header, and check that any other line contains something that conforms to a URL regex of at least example.com (and probably be whitespace-lenient, just strip it before match or ignore in the regex).

**Comment by anusharanganathan** [comment]
1. Going with whitelist file being in YAML format. I can check for conformity easily. It is easy for humans too. 
2. I can either validate the string to be a valid URL (using [urlparse](https://docs.python.org/2/library/urlparse.html)) or validate a string to be a valid regex - not sure how I could validate a regex url string to be a valid url. Also, would prefer to use url parse rather than regex for validating a URL. Looking at the example whitelist given, urlparse would work.
```
# Validate regex
import re
try:
    re.compile('[')
    is_valid = True
except re.error:
    is_valid = False
```
I too think maintaining a canonical version outside of the server would be better and if a file is posted, we expect that to be the current canonical version. No CRUD options for single urls. If it is needed, we would need to put some thought into how we address each url and yes, how we manage bulk updates.

We could just hold all versions of the whitelist files they post to us (or the last x versions, if it is being constantly updated) - so they can retrieve a previous version if they so wish. The version in use would be the one last posted.

**Comment by markmacgillivray** [comment]
I don't think yaml is easy enough, actually, particularly as it requires whitespace formatting to be present which is not at all obvious to non-technical people. Is there any problem with just using a simple approach of comments / headers being prepended with a #, and any other line being a valid URL? Valid as in parses as a URL, not necessarily resolvable.

See the issue on the chrome plugin side of things too, and talk to @Steven-Eardley about it if you have not - the only real technical requirement is that he can have the plugin read the whitelist well enough. See issue: https://github.com/CottageLabs/webber_chrome/issues/8

**Comment by anusharanganathan** [comment]
@markmacgillivray  @Steven-Eardley  Okay yes, YAML needs to be white space formatted, but it handles comments, sections headings, spaces, empty lines. I was just thinking of not re-inventing the wheel. If we just go with a text file, we still need rules
  * Lines that are comments start with a #
  * Section headings should be in square brackets (I suppose)
  * For the sake of simplicity, lets just say no inline comments (i.e.) url # this is a comment

Well, I guess custom rules in a text file it is.

**Comment by Steven-Eardley** [comment]
Aren't headings just for human convenience? They could just be simple comments too, in that case.

**Comment by markmacgillivray** [comment]
Yep, all we need is # for lines that are not URLs. They can be comments or
headers. Every other line is either blank (or has some random amount of
whitespace in it, which we will count as blank) or validates as a URL.

**Comment by anusharanganathan** [comment]
Oh could be. Would they want us to save information from the pages differently, based on sections, in the future? If so, we need to parse the sections.

**Comment by markmacgillivray** [comment]
Also be lenient with URL matches - do not expect http or even www. Probably
the strongest match to make is that it is at least example.com

**Comment by markmacgillivray** [comment]
So far headers are just human readable information, so work on that basis
for now.

**Comment by anusharanganathan** [comment]
Yes, I will make the url matching pretty lenient

**Comment by bodobalazs** [comment]
you can safely assume that people who will edit the whitelist file will have programming experience. nevertheless, no whitelist should go live without validation.

**Comment by anusharanganathan** [comment]
@bodobalazs could you please elaborate on the validation. Did you have a process in mind for the validation?
